packages/core/src/components/loading-skeleton/loadingSkeleton.tsx

```diff
@@ -0,0 +1,136 @@
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as classNames from "classnames";
+import * as React from "react";
+
+import { AbstractComponent, Classes, IProps } from "../../common";
+
+export interface ILoadingSkeletonProps extends IProps {
+    /**
+     * If true, show an animated loading skeleton when `isLoading` is true. Otherwise show a static
+     * skeleton.
+     * @default true
+     */
+    animated?: boolean;
+
+    /**
+     * If true, show a loading skeleton. Otherwise render this component's child.
+     */
+    isLoading: boolean;
+
+    /**
+     * Specifies the number of animated lines to show.
+     * @default 1
+     */
+    numBones?: number;
+
+    /**
+     * If true, render a random width loading skeleton. This can be used to make things like table
+     * cells display a more natural-feeling loading state.
+     * @default false
+     */
+    randomWidth?: boolean;
+}
+
+export interface ILoadingSkeletonState {
+    animated?: boolean;
+    rightMargins?: number[];
+}
+
+export class LoadingSkeleton extends AbstractComponent<ILoadingSkeletonProps, ILoadingSkeletonState> {
+    public static defaultProps: ILoadingSkeletonProps = {
+        animated: true,
+        isLoading: true,
+        numBones: 1,
+        randomWidth: false,
+    };
+
+    public state: ILoadingSkeletonState;
+
+    private boneRefHandlers: Array<((ref: HTMLDivElement) => void)> = [];
+    private boneRefs: HTMLDivElement[];
+
+    public constructor(props: ILoadingSkeletonProps, context?: any) {
+        super(props, context);
+        const { numBones, randomWidth } = props;
+        const rightMargins: number[] = [];
+        for (let i = 0; i < numBones; i++) {
+            rightMargins.push(this.generateRightMargin(randomWidth));
+        }
+        this.boneRefs = [];
+        this.state = { animated: true, rightMargins };
+    }
+
+    public componentWillReceiveProps(nextProps: ILoadingSkeletonProps) {
+        const { numBones, randomWidth } = this.props;
+        const { animated: nextAnimated, numBones: nextNumBones, randomWidth: nextRandomWidth } = nextProps;
+        let rightMargins = this.state.rightMargins.slice();
+
+        if (nextRandomWidth !== randomWidth) {
+            rightMargins = [];
+            for (let i = 0; i < nextNumBones; i++) {
+                rightMargins.push(this.generateRightMargin(nextRandomWidth));
+            }
+            this.setState({ rightMargins });
+        } else if (nextNumBones > numBones) {
+            if (rightMargins.length !== nextNumBones) {
+                for (let i = numBones; i < nextNumBones; i++) {
+                    rightMargins.push(this.generateRightMargin(randomWidth));
+                }
+            }
+
+            // sync animations
+            this.setState({ rightMargins });
```

Comment:
this is handling for a rather specific case where you're adding more bones after the component has already mounted. the new bones start animating without regard to the old bone animations.